### ISSUE TITLE
Presubmit tests for Windows in 1.20 and release blocking tests

### ISSUE THREAD

**Issue by jsturtevant** [kubernetes/kubernetes#93276]
**What would you like to be added**:
Presubmit tests for windows  in 1.20 and to see windows tests in release blocking [dashboard like 1.19](https://k8s-testgrid.appspot.com/sig-release-1.19-blocking) for 1.20

**Why is this needed**:
The current process for testing changes against Windows and identifying issues is reactive which causes lots of work after the fact.  For 1.19 we had at least 2 PR's that introduced breaking changes after code freeze. I believe this has happened in prior releases as well. 

Two bugs that caused windows nodes to not work recently:

- https://github.com/kubernetes/kubernetes/pull/80917
- https://github.com/kubernetes/kubernetes/pull/90948

The result was 1.19rc.1 did not work on windows and many tests built of main branch for [sig-windows](https://testgrid.k8s.io/sig-windows) tests failed for over a week.  Both issues needed to be fixed via a last minute pr for 1.19rc2: https://github.com/kubernetes/kubernetes/pull/93263

## Next Steps
What are the requirements for adding a presubmitts from the tests?  Is this documented somewhere (I did some searching and wasn't able to find anything)?

A few thoughts to start:

- [ ] Need to show stability in tests (can this be a subset of tests to start?)
- [ ] Speed of tests should be consistently under 2hrs (based on https://k8s-testgrid.appspot.com/sig-release-master-blocking#Conformance%20-%20GCE%20-%20master)
- [ ] need sig-windows support to drive out flakes that might be introduced
- [ ] can there be a subset of tests that can be run to atleast validate windows nodes come up and simple pods are scheduled to start
- [ ] do the blocking PR's need to be run on GCE or can we use tests from https://testgrid.k8s.io/sig-windows#aks-engine-azure-windows-master-staging?
- [ ] Image builder prs for windows need to be merged: https://github.com/kubernetes/kubernetes/pull/77398 and https://github.com/kubernetes/kubernetes/pull/77269 this will resolve issues when the images are changed and not require manual builds and pushes are required now.

My goal here is to learn what is required and start the conversation.  I am happy to help drive the conversation in the correct places.

/sig windows
/sig testing 
/sig release

**Comment by liggitt** [comment]
cc @BenTheElder @kubernetes/sig-testing 

> Speed of tests should be consistently under 2hrs

Presubmits should be **way** less than 2 hours... ideally in the 10-30 minute range. The presubmits we have that are in the 1+ hour range are considered problems... let's not add more like that.

**Comment by BenTheElder** [comment]
Responding to some of this in depth, happy to discuss more later.

----

Release candidates not working and this being caught is actually working as intended IMHO, thanks for reporting these issues.

The goal of the release candidates is for people to try them and report issues like this before the actual release.

> What are the requirements for adding a presubmitts from the tests? Is this documented somewhere (I did some searching and wasn't able to find anything)?

There aren't any written currently because it's not something we do frequently and it's generally discouraged.
There are plenty of other avenues for test signal that ensure the quality of releases while consuming much less resources and avoid limiting the merge rate.

We don't even do a real cross compile in presubmit (we x-typecheck the code instead) because building for all of these targets is excessively expensive.

The current trend / intention is to move more things _out_ of presubmit.

There are however lots of _optional_, _manually triggered_ presubmits that various SIGs may use to test their changes.
Reviewers & approvers maintaining e.g. windows specific code are welcome to run more tests and check their results before approving changes.

For release blocking there is documentation.

https://github.com/kubernetes/sig-release/blob/master/release-blocking-jobs.md#release-blocking-jobs-and-criteria

> Presubmits should be way less than 2 hours... ideally in the 10-30 minute range. The presubmits we have that are in the 1+ hour range are considered problems... let's not add more like that.

💯 
Even the 1 hour stuff is a poor experience and seriously limits merge rates / retest. We've long intended for things to be < 30m. 2 hours is not acceptable for presubmit.

>  do the blocking PR's need to be run on GCE or can we use tests from https://testgrid.k8s.io/sig-windows#aks-engine-azure-windows-master-staging?

They need to run on Prow (the community CI) to be in presubmit.
Some of them don't even use cloud, and usage of other clouds is fine IF appropriate resources are provided.

However SIG Testing is also not in favor of cloud in presubmit (as opposed to postubmit / periodic tests monitored by other SIGs and the release team) and we've been making strides towards removing the existing cloudy bits as we'd prefer presubmits to be:
- runnable by any member of the community (I.E. not dependent on cloud privilege)
- fast
- not severely restricted in concurrency (I.E. we may run many PRs at once)
- not dependent on infrastructure the community / staffed oncall can't fix ourselves

Other tests are fine to run outside of PR blocking.

>  Image builder prs for windows need to be merged: #77398 and #77269 this will resolve issues when the images are changed and not require manual builds and pushes are required now.

We need to have a discussion at some point about our test image builds depending on magic VMs the community / CNCF does not own. This was a major regression that I wasn't initially aware of.

**Comment by BenTheElder** [comment]
I also want to note that we've had existing open issues about Windows CI for a very long time and not seem much work done on them.

I'd kinda like to see reliable CI in release informing before further discussion about release blocking or presubmit blocking.

Issues like https://github.com/kubernetes/kubernetes/issues/93278 have been very common so far, and often subtly hidden unless you inspect the results in depth.

**Comment by jsturtevant** [comment]
@liggitt and @BenTheElder thanks for the quick and detailed responses! This helps with context and understanding.  I have a few additional comments/questions:

> We don't even do a real cross compile in presubmit (we x-typecheck the code instead) because building for all of these targets is excessively expensive.

When it comes to issues like (https://github.com/kubernetes/kubernetes/pull/90948) which don't touch windows but still break it,  would like to think about an solution that would help catch breaking changes.  This is were I think a presubmit could come into play but as you pointed out we need to create a reliable signal in CI first.  Without some check from a windows perspective in a presubmit we end up with tests broken for a long time period (was a week in this case) and another PR broke it further during this time.

From this comment, sounds like with current set up, might not be able to do windows checks since it's not cross-compiled?

> There are however lots of optional, manually triggered presubmits that various SIGs may use to test their changes.
Reviewers & approvers maintaining e.g. windows specific code are welcome to run more tests and check their results before approving changes.

I think as a sig we need to agree on a subset of these and start using them regularly, which would be a good start. I recently [re-organized the test-grid](https://github.com/kubernetes/test-infra/pull/18123) so it's a bit easier to see the [presubmits](https://testgrid.k8s.io/sig-windows-presubmit) and can start to drive down the flakes.  #80917 should have been caught by a test like that.

@michmike maybe we can discuss this at our next sig-windows meeting?

> we'd prefer presubmits to be:
    runnable by any member of the community (I.E. not dependent on cloud privilege)
    fast
    not severely restricted in concurrency (I.E. we may run many PRs at once)
    not dependent on infrastructure the community / staffed oncall can't fix ourselves

It looks like `kind` is one of the avenues for this? I see there is an open issue to work on windows support which would potentially open the door to this  (https://github.com/kubernetes-sigs/kind/issues/1529)

> I'd kinda like to see reliable CI in release informing before further discussion about release blocking or presubmit blocking.

Acknowledged. I think this is the first actionable item which we can work towards.

**Comment by jsturtevant** [comment]
>We need to have a discussion at some point about our test image builds depending on magic VMs the community / CNCF does not own. This was a major regression that I wasn't initially aware of.

My understanding is these prs should resolve that issue?  When the PR's get merged the support for building windows images should be in the CNCF?  cc: @claudiubelu

**Comment by BenTheElder** [comment]
> When it comes to issues like (#90948) which don't touch windows but still break it, would like to think about an solution that would help catch breaking changes. This is were I think a presubmit could come into play but as you pointed out we need to create a reliable signal in CI first. Without some check from a windows perspective in a presubmit we end up with tests broken for a long time period (was a week in this case) and another PR broke it further during this time.

We can catch windows breakage and fix it without a presubmit. E.G. many alpha and beta-grade features do this.
We have configurable support for sending prompt alert emails when a test transitions to failing.

The fact that it was broken for a week tells me that we currently have one or all of:
- Windows CI is too slow
- Windows CI is poorly monitored
- Windows fixes take too long to develop
- It takes too long to merge fixes

All of which are resolveable without a presubmit. The last of which is helped by not introducing new flake sources or slow tests.

This is also the case for many things like e.g 5k node scale testing. Not everything is in presubmit.

> From this comment, sounds like with current set up, might not be able to do windows checks since it's not cross-compiled?

It's worse than that, there's no pipelining in CI and no ETA on support for that, so every presubmit we always run means another build we must always do concurrently. Adding another always-on presubmit means adding another concurrent build of everything.

We already run 9 e2e presubmit (including node e2e) and a plain build-everything on every PR, every push, which is ... a bit much (not to mention non-build/e2e tests like linting).

This is _somewhat_ alleviated by all being for the same platform and sharing a bazel remote build cache.

Periodic testing does not suffer from this, we continuously publish public pre-compiled artifacts which the e2e tests simply download and run instead of building.

Resource contention is a major suspect in #92937 

> It looks like kind is one of the avenues for this? I see there is an open issue to work on windows support which would potentially open the door to this (kubernetes-sigs/kind#1529)

No, Windows containers cannot run a container runtime inside them, and prow does not have windows nodes. This is infeasible. You also cannot run a windows cluster without a linux control plane, so you'd need a runtime that could have linux containers alongside windows containers even if the windows container nodes were possible.

kubernetes-sigs/kind#1529 is about having CI signal that linux clusters work correctly on windows hosts. 
Currently we don't have any known CI platforms where this is possible to test. 
We need to create linux docker containers on windows.

> My understanding is these prs should resolve that issue? When the PR's get merged the support for building windows images should be in the CNCF? cc: @claudiubelu

That is not my understanding. Those PRs do not relate to the VMs, which AIUI are not owned by the CNCF / wg k8s.
We have legacy google-owned resources, but that has been actively transitioned off-of, and no new infrastructure like this has been created in O(years).
Certainly not hard requirements for building artifacts we publish.

**Comment by claudiubelu** [comment]
> 
> > My understanding is these prs should resolve that issue? When the PR's get merged the support for building windows images should be in the CNCF? cc: @claudiubelu
> 
> That is not my understanding. Those PRs do not relate to the VMs, which AIUI are not owned by the CNCF / wg k8s.

So, there's a longer story there.

First of all, the Image Promoter Windows build nodes are on the CNCF azure subscription that is owned by Microsoft / sig-windows. They are dedicated nodes meant for building those images.

Secondly, that was the compromise we've reached long ago. The original plan was to have google spin up those Windows VMs required by the Image Promoter. We've even had to make some changes to take into account the fact that Google does not support Hyper-V isolation, so we couldn't rely on it to build Windows Container images for multiple OS versions.

But in the end, there was no progress on the Google side. So, we've offered our own resources / nodes for it instead. Sure, there were a few hiccups on the configuration side, but they were taken care of (e.g.: getting those necesarry ceritificates to the GCB instance itself).

Now, if that's still undesirable for you, we could try minimizing the need for Windows nodes. We could use ``docker buildx`` for most of the building process, although that would tie our hands behind our backs for the most part because it would basically mean we can no longer ``RUN`` things in our Dockerfiles.

But, we can have a workaround for that: We can build a Windows helper image, built on a Windows Server node (we only need one, not multiple) that can solve most / all of those `RUN` commands, and then we can just copy over the needed files in ``docker buildx`` which will run on Linux entirely. Now, here's a question: do you want to build this image every time, or just have it hosted on ``e2team/windows-image-builder-helper:1.0`` or something similar? IMO, the first option would be preferable; if things need to be changed about it, or have new things installed, other people can easily do it themselves, but that means that you'll still have to keep 1 "magic VM", as you say.

And lastly, there's one more important bit: Windows container images are huge (servercore images). Which is why we've switched from servercore-based images to nanoserver-based images, which are ~10x smaller. But, the nanoserver images still require some DLLs, so we fetch them from their servercore equivalent images. Just pulling and unpacking those servercore images could take 20 minutes. So, I was thinking about having them somewhere on github (somewhere in kubernetes-sig I think) and getting them from there, or we could have a second helper image: a Linux scratch image which will contain those necessary DLLs, which will be rebuilt weekly or so.

Another benefit for the buildx implementation would be that it'll be easier to add the WS 20.04 images; we wouldn't have create a new node for it, assign public IP / DNS name, regenerate certificates to also include it, have someone update them in prow. Same for future releases too.

This is how the changes would look (keep in mind: WIP POC): https://github.com/claudiubelu/kubernetes/commit/eb70e48f222348cba2234f6e560ebf8c65431825

**Comment by BenTheElder** [comment]
> Now, if that's still undesirable for you, we could try minimizing the need for Windows nodes. We could use docker buildx for most of the building process, although that would tie our hands behind our backs for the most part because it would basically mean we can no longer RUN things in our Dockerfiles.

The most undersirable part is using VMs that are not owned / managed by the project at large. Anywhere we're doing that is infrastructure debt that the community can't work on / has no visibility into / ability to fix.

GCP _and_ AWS have resources provided to the CNCF managed via wg-k8s-infra that the community is moving the project onto.
You can find the configuration for these wholly in https://github.com/kubernetes/k8s.io, nothing is hand managed by a particular vendor, instead by the work group, even the ACLs for which members have direct access to these resources are synced out from the repo.

Not requiring remote VMs would of course be even better, so anyone can reproduce / build them locally. 

Distinct from the single-company managed resources which we should avoid introducing more of, it's somewhat problematic that contributors can't build these outside of CI, I don't think we have any other images where this is the case.

Many of the builds do use for example, the project's Google Cloud Build resources but they don't fundamentally depend on it, and the build specs can be run locally.

> Another benefit for the buildx implementation would be that it'll be easier to add the WS 20.04 images; we wouldn't have create a new node for it, assign public IP / DNS name, regenerate certificates to also include it, have someone update them in prow. Same for future releases too.

This would also be a great benefit.

**Comment by jsturtevant** [comment]
@claudiubelu Thanks for the details on the current Image Promoter status.  Hope we can come to agreement on the buildx support reducing the need for VMs and making it easier to reproduce and build locally.

Going to try to summarize the long thread:

- true cross builds do not happen for pre-submits due to high resource requirements. (issue tracking this?)
- there is a desire from sig-testing to lower the pre-submit build time (< 30m is long term goal) and remove cloudy components (to support local runs by all member of the community without need to have access to cloud resources)
- there is a trend to move things out of pre-submitts but sigs are encourage to create optional, manually triggered presubmits to test code added
- tests need to be reliable in the CI in release informing before further discussion about release blocking or presubmit blocking.

and some proposed actions items for would be:

- [ ] [sig-windows] agree upon a set of pre-submits that should run before windows related changes and demonstrate stability. This can be discussed at sig-windows. Some options are:
    - Since going beta in 1.19 the [containerd builds](https://testgrid.k8s.io/sig-windows-containerd#aks-engine-azure-windows-master-containerd) have been stable (minus the e2e images issues due to image promoter status) and we would like to move containerd to stable with 1.20.  Using these would be useful signal.
    - The [azure disk PR](https://testgrid.k8s.io/sig-windows-presubmit#pr-k8s-azure-disk-e2e-master-windows) runs currently any windows file is touched. This was introduced because there had been regression in the past for disks.
    - Recently [deflacked](https://github.com/kubernetes/kubernetes/pull/93588) [Azure staging](https://testgrid.k8s.io/sig-windows-releases#aks-engine-azure-windows-master-staging)
    -  [gce-windows staging](https://testgrid.k8s.io/sig-windows-releases#gce-windows-2019-master)
- [ ] [sig-windows] develop a e2e test that can be run without cloud support. We don't currently have this as far as I know.  [kind](https://kind.sigs.k8s.io/) is not an option (can't run containers in containers) but [minikube could be](https://github.com/kubernetes/minikube/issues/2015).  Are there other options?
- [ ] [sig-windows/sig-testing]Merge windows support to Image Promoter.  Without Image Promoter support the e2e tests will always have failures when images are updated causing reliability as well as difficulty building locally.
- [ ] [sig-windows] Demonstrate stability in release-informing CI

There is still a desire to test windows at a pre-submit to catch issues like #90948 where changes to Linux behavior caused breaking changes on Windows. As these action items complete we can readdress Release blocking and pre-submits.  

@michmike @marosset @adelina-t @pjh @ddebroy thoughts/comments from sig-windows?

**Comment by immuzz** [comment]
SIG Windows Backlog Meeting 

@jsturtevant to bring this for discussion in next SIG Windows Meeting

**Comment by BenTheElder** [comment]
>  [sig-windows] develop a e2e test that can be run without cloud support. We don't currently have this as far as I know. kind is not an option (can't run containers in containers) but minikube could be. Are there other options?

We run our CI in containers, which is why kind exists to begin with (because on linux we can run containers in containers), we can't run VMs there.

>  [sig-windows] agree upon a set of pre-submits that should run before windows related changes and demonstrate stability. This can be discussed at sig-windows. Some options are:

I ask to please focus on getting stable release blocking signal that release-ci-signal is happy with first, then we can explore presubmit.

**Comment by marosset** [comment]
> GCP _and_ AWS have resources provided to the CNCF managed via wg-k8s-infra that the community is moving the project onto.

Azure provides dedicated resources to the CNCF as well. 
All of the periodic jobs that test Windows, cloud-provider-azure, CAPZ, run on resources in here and we've also given resources in here to other projects like minikube for some of their CI testing. 

@BenTheElder can we work with wg-k8s-infra to improve the ACL story/management of for these resources to be more community focused?

**Comment by BenTheElder** [comment]
Resources managed by wg-k8s-infra are fully controlled by members from multiple employers across the wg with only the billing handled via CNCF / provider, unblocking community management and insight into resource allocation. Please do work with the group. Configs for everything are here https://github.com/kubernetes/k8s.io

**Comment by jsturtevant** [comment]
> I ask to please focus on getting stable release blocking signal that release-ci-signal is happy with first, then we can explore presubmit.

@BenTheElder I was referring to the "optional, manually triggered presubmits" that was a suggestion from a previous comment.  I think it would be valuable for sig-windows to begin to run subset of optional tests that are agreed upon for windows changes to attempt to catch any regressions before they hit the CI tests.

I agree we need to focus on the stability of the release-ci, there are currently two informing tests, which are in better state than they were during the 1.19 release (when there was steadily failing tests), so have had some improvement.

**Comment by ravisantoshgudimetla** [comment]
Update: On the sig-windows side, we are trying to improve the stability of serial tests and we're monitoring the release blocking signals. Hopefully in the next few weeks we will get a solid green on the release blocking signals. If we look at the https://testgrid.k8s.io/sig-windows-releases#gce-windows-2019-master except for a couple of storage tests, we will try to deflake them in the next few weeks and make this test mandatory for presubmits

**Comment by jsturtevant** [comment]
Linking to the pre-submit cross compile issue: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/bazel.md#cross-compilation-of-cgo-is-not-currently-natively-supported

**Comment by ravisantoshgudimetla** [comment]
cc @dims

**Comment by jsturtevant** [comment]
We discussed this at length today in various forms. Here is an attempt at the summary:

- We should be able to get the signal we are looking for by adding post submits and periodic's for lower costs (monetary and operational)
- Generally, direction is move things out of pre-submit but if we can show it's necessary we can pick up the thread on how to run Windows Pre-submits. 
- The ideal testing graduation flow at this time would be: Informing -> Release Blocking and **_If_** deemed necessary after watching the signal in Release Blocking and see that there are issues causing breaking changes regularly we could revisit.

Next steps:

- Turn informing jobs to stable green signal so we can move them towards Release blocking. The release blocking criteria: https://github.com/kubernetes/sig-release/blob/master/release-blocking-jobs.md#release-blocking-criteria-and-dashboard
- PR for adding containerd release informing: https://github.com/kubernetes/test-infra/pull/20525
- Complete our transition for image promotion (several test images are now promoted with several open prs for the remaining)

@ravisantoshgudimetla @jayunit100 please feel free to add anything I missed or doesn't sound right

**Comment by dims** [comment]
yay bazel is gone, so this is off my lap

/unassign
